sys/dev/isp/isp.c

```diff
@@ -4322,21 +4338,341 @@ isp_reinit(ispsoftc_t *isp, int do_load_defaults)
+	    (fcp->flt_length > (FLT_HEADER_SIZE + FLT_REGIONS_SIZE))) {
+		isp_prt(isp, ISP_LOGERR,
+		    "FLT[DEF]: Invalid length=0x%x(%d)",
+		    fcp->flt_length, fcp->flt_length);
+	}
+	isp_prt(isp, ISP_LOGDEBUG0,
+	    "FLT[DEF]: version=0x%x length=0x%x(%d) checksum=0x%x",
+	    ver, fcp->flt_length, fcp->flt_length, csum);
+}
+
+static int
+isp_read_flt_2xxx(ispsoftc_t *isp)
+{
+	fcparam *fcp = FCPARAM(isp, 0);
+	int retval = 0;
+	int len = fcp->flt_length - FLT_HEADER_SIZE;
+	uint32_t addr, lwrds, *dptr;
+	uint8_t flt_data[len];
+	fcp->flt_region_entries = len / FLT_REGION_SIZE;
+
+	addr = fcp->flt_region_flt + (FLT_HEADER_SIZE >> 2);
+	dptr = (uint32_t *) flt_data;
+	isp_prt(isp, ISP_LOGDEBUG0, "FLT[DEF]: regions=%d",
+	    fcp->flt_region_entries);
+	for (lwrds = 0; lwrds < len >> 2; lwrds++) {
+		isp_rd_2xxx_flash(isp, addr++, dptr++);
+	}
+	retval = isp_parse_flt_2xxx(isp, flt_data);
+	return (retval);
+}
+
+static int
+isp_parse_flt_2xxx(ispsoftc_t *isp, uint8_t *flt_data)
+{
+	fcparam *fcp = FCPARAM(isp, 0);
+	int count;
+	struct flt_region region[fcp->flt_region_entries];
+
+	for (count = 0; count < fcp->flt_region_entries; count++) {
+		region[count].code =
+		    le16toh((uint16_t) (ISP2XXX_FLT_REG_CODE(flt_data, count)));
+		region[count].attribute =
+		    (uint8_t) (ISP2XXX_FLT_REG_ATTR(flt_data, count));
+		region[count].reserved =
+		    (uint8_t) (ISP2XXX_FLT_REG_RES(flt_data, count));
+		region[count].size =
+		    le32toh((uint32_t) (ISP2XXX_FLT_REG_SIZE(flt_data, count)) >> 2);
+		region[count].start =
+		    le32toh((uint32_t) (ISP2XXX_FLT_REG_START(flt_data, count)) >> 2);
+		region[count].end =
+		    le32toh((uint32_t) (ISP2XXX_FLT_REG_END(flt_data, count)) >> 2);
+
+		isp_prt(isp, ISP_LOGDEBUG0,
+		    "FLT[0x%x]: start=0x%x end=0x%x size=0x%x attribute=0x%x",
+		    region[count].code, region[count].start, region[count].end,
+		    region[count].size, region[count].attribute);
+
+		switch(region[count].code) {
+		case FLT_REG_FW:
+			fcp->flt_region_fw = region[count].start;
+			break;
+		case FLT_REG_BOOT_CODE:
+			fcp->flt_region_boot = region[count].start;
+			break;
+		case FLT_REG_VPD_0:
+			fcp->flt_region_vpd_nvram = region[count].start;
+			if (isp->isp_port == 0)
+				fcp->flt_region_vpd = region[count].start;
+			break;
+		case FLT_REG_VPD_1:
+			if (isp->isp_port == 1)
+				fcp->flt_region_vpd = region[count].start;
+			break;
+		case FLT_REG_VPD_2:
+			if (!IS_27XX(isp)) 
+				break;
+			if (isp->isp_port == 2)
+				fcp->flt_region_vpd = region[count].start;
+			break;
+		case FLT_REG_VPD_3:
+			if (!IS_27XX(isp)) 
+				break;
+			if (isp->isp_port == 3)
+				fcp->flt_region_vpd = region[count].start;
+			break;
+		case FLT_REG_NVRAM_0:
+			if (isp->isp_port == 0)
+				fcp->flt_region_nvram = region[count].start;
+			break;
+		case FLT_REG_NVRAM_1:
+			if (isp->isp_port == 1)
+				fcp->flt_region_nvram = region[count].start;
+			break;
+		case FLT_REG_NVRAM_2:
+			if (!IS_27XX(isp)) 
+				break;
```

Comment:
I wonder where quad-port pre-27xx cards will take nvram addresses for the last two ports?  Or if they are supposed to fall back to hardcoded addresses, then how?

Reply:
Good catch, I will take a look at how this is done in the official Linux driver.

Reply:
As far as I can tell from the official Linux driver there shouldn't be any pre-27xx cards? At least I can't find anything about it in the code.

Details:

The original Linux driver code for the third and fourth port is:
```
case FLT_REG_NVRAM_2:
        if (!IS_QLA27XX(ha) && !IS_QLA28XX(ha))
                break;
```
This resolves in the end to the following PCI IDs (stripped down a bit to keep it shorter):
```
#define PCI_DEVICE_ID_QLOGIC_ISP2071    0x2071
#define PCI_DEVICE_ID_QLOGIC_ISP2271    0x2271
#define PCI_DEVICE_ID_QLOGIC_ISP2261    0x2261
#define PCI_DEVICE_ID_QLOGIC_ISP2081    0x2081
#define PCI_DEVICE_ID_QLOGIC_ISP2281    0x2281
#define PCI_DEVICE_ID_QLOGIC_ISP2971    0x2971

#define IS_QLA2071(ha)  (DT_MASK(ha) & DT_ISP2071)
#define IS_QLA2271(ha)  (DT_MASK(ha) & DT_ISP2271)
#define IS_QLA2261(ha)  (DT_MASK(ha) & DT_ISP2261)
#define IS_QLA2971(ha)  (DT_MASK(ha) & DT_ISP2971)
#define IS_QLA2081(ha)  (DT_MASK(ha) & DT_ISP2081)
#define IS_QLA2281(ha)  (DT_MASK(ha) & DT_ISP2281)

#define IS_QLA27XX(ha)  (IS_QLA2071(ha) || IS_QLA2271(ha) || IS_QLA2261(ha) || \
                        IS_QLA2971(ha))
#define IS_QLA28XX(ha)  (IS_QLA2081(ha) || IS_QLA2281(ha))
```

Compared to the following code in our driver  (stripped down a bit to keep it shorter):
```
#define PCI_PRODUCT_QLOGIC_ISP2684      0x2171
#define PCI_PRODUCT_QLOGIC_ISP2692      0x2b61
#define PCI_PRODUCT_QLOGIC_ISP2714      0x2071
/* 0x2271 not mentioned/supported at all */
#define PCI_PRODUCT_QLOGIC_ISP2722      0x2261
#define PCI_PRODUCT_QLOGIC_ISP2814      0x2081 /* new with my changes */
#define PCI_PRODUCT_QLOGIC_ISP2812      0x2281 /* new with my changes */
/* 0x2971 not mentioned/supported at all */

case PCI_QLOGIC_ISP2684:
case PCI_QLOGIC_ISP2692:
case PCI_QLOGIC_ISP2714:
case PCI_QLOGIC_ISP2722:
        did = 0x2700;
        isp->isp_mdvec = &mdvec_2700;
        isp->isp_type = ISP_HA_FC_2700;
        break;
case PCI_QLOGIC_ISP2812:
case PCI_QLOGIC_ISP2814:
        did = 0x2800;
        isp->isp_mdvec = &mdvec_2800;
        isp->isp_type = ISP_HA_FC_2800;
        break;

#define IS_27XX(isp)    ((isp)->isp_type >= ISP_HA_FC_2700)
#define IS_28XX(isp)    ((isp)->isp_type >= ISP_HA_FC_2800)
```

So from this point of view it is quite the same.
Looking at the debug output of my 25xx based 2532 dual port card the FLT contains no NVRAM info about the third or fourth port whereas the FLTs of my 27xx based 2722 single port card and my 28xx based 2812 dual port card both contain NVRAM info for all four ports.

I couldn't find anything useful about pre-27xx four port cards. Do you have something or are you aware of such cards?

If you are in doubt about it we can just remove the
```
if (!IS_27XX(isp)) 
	break;
```
for NVRAM_2 and NVRAM_3?

_Strange_ fact - the following PCI IDs we have as 27xx based devices
```
#define PCI_PRODUCT_QLOGIC_ISP2684      0x2171
#define PCI_PRODUCT_QLOGIC_ISP2692      0x2b61
```
don't show up in the official Linux driver at all and I was unable to find some useful information about it.

Reply:
I could be wrong.  It may be that quad-port cards before certain point (pre-27xx) were dual-chip.  So it may be not a problem indeed.

Reply:
QLE2564 and QLE2664 seem to be dual-chip, so it may have sense.